### PR TITLE
Transform pixel track fitter, filter, and cleaner to ED or ES products

### PR DESCRIPTION
**Pull request by makortel**
This PR contains intermediate developments before migrating `PixelTrackProducer` to the new seeding framework of #16635. The point is to make all other components than the hit ntuplet generation to support `fillDescriptions()`. This is achieved by making fitters and filters EDProducts (because at least one of the algorithms depend on event data) and making cleaners ESProducts (because none of the cleaners need event data).

I have tried to migrate all uses in CMSSW, but I can't exclude of missing something.

The next step is to migrate `PixelTrackProducer` to the new seeding framework (#16635), and fully migrate HLT (with `customizeHLTforCMSSW.py`).

Tested in CMSSW_9_0_X_2016-11-27-2300 (rebased on top of CMSSW_9_0_X_2016-12-04-2300), no changes expected.

@rovere @VinInn @felicepantaleo @mtosi @JanFSchulte

### CODE REVIEW

**Comment by makortel** [comment]
@cmsbuild, please test

**Comment by makortel** [comment]
@cmsbuild, please test

(trivial typo; but are there instructions somewhere how I could run the "AddOn" tests myself?)

**Comment by Martin-Grunewald** [comment]
```addOnTests.py```

**Comment by makortel** [comment]
@Martin-Grunewald Thanks, very simple.

@smuzaffar Would it make sense to give the recipe to re-run the AddOn tests in the cms-bot message in case the tests failed (or in the test result page)?

**Comment by smuzaffar** [comment]
@makortel , I will update cms-bot to add the recipe to run addon test.

**Comment by makortel** [comment]
@smuzaffar Thanks!

File: HLTrigger/Configuration/python/customizeHLTforCMSSW.py [hunk 1]
```diff
@@ -90,6 +168,7 @@ def customizeHLTforCMSSW(process, menuType="GRun"):
         process = customiseFor15440(process)
         process = customiseFor15499(process)
         process = customiseFor16569(process)
+        process = customiseFor16792(process)
```
Comment (Martin-Grunewald) [inline HLTrigger/Configuration/python/customizeHLTforCMSSW.py]: Yes, please.

**Comment by makortel** [comment]
@cmsbuild, please test

Moved the HLT customize function call under `if cmsswVersion >= "CMSSW_9_0":`, and added customize code to remove `useFilterWithES` parameter from `PixelTrackProducer` (it became obsolete; removing it now means that I don't have to deal with it in the "next PR" where I'll add `fillDescriptions()` to `PixelTrackProducer`).

**Comment by makortel** [comment]
@cmsbuild, please test

Rebased on top of CMSSW_9_0_X_2016-12-04-2300.

**Comment by makortel** [comment]
@slava77 Regarding your question if this PR would conflict with #16885 (it was that, right?), the answer is **no**. They modify different files.

**Comment by slava77** [comment]
On 12/8/16 4:34 AM, Matti Kortelainen wrote:
> @slava77 <https://github.com/slava77> Regarding your question if this PR
> would conflict with #16885 <https://github.com/cms-sw/cmssw/pull/16885>
> (it was that, right?), the answer is *no*. They modify different files.
> 

Thank you for the confirmation.

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/16792 1cf669c
- mostly technical changes moving filter/fitter/cleaner tools internal to pixel track producer  modules  to ES or ED products
- jenkins tests pass and comparisons with baseline show no relevant differences
    - some HLT HeavyFlavor Upsilon plots (e.g. /HLT_Mu7p5_Track3p5_Upsilon_v) have changed in a way that suggests some random issues, like the same histogram written to by different modules

**Comment by makortel** [comment]
> some HLT HeavyFlavor Upsilon plots (e.g. /HLT_Mu7p5_Track3p5_Upsilon_v) have changed in a way that suggests some random issues, like the same histogram written to by different modules

Do we have a list of "known" DQM histograms with reproducibility issues? If not could we make one (and maybe get the issues fixed at some point)? If @slava77 is referring to e.g. 25202.0, I've just wasted some hours and hairs with this histogram when developing the next step after this PR.

**Comment by slava77** [comment]
We regularly ignore the following pattern

/HLT_IsoTkMu\|/HLT_OldIsoTkMu

The Upsilon plot hasn't come up before,
however, given that the title of the histogram changed, I noted that it's something random (diff is done based on the histogram name)

wf25202_upsilon_hlt

**Comment by makortel** [comment]
Actually this might be related to this PR. `hltPixelTracksL3MuonFilter` is a module added in `customizeHLTforCMSSW` of type `PixelTrackFilterByKinematicsProducer`, whereas `hltMu7p5Track2UpsilonTrackMassFiltered` is of type `HLTMuonTrackMassFilter`. I wonder how this change enters up in a DQM histogram though. Ok, taking a quick look on the code, this logic
https://github.com/cms-sw/cmssw/blob/CMSSW_9_0_X/HLTriggerOffline/HeavyFlavor/src/HeavyFlavorValidation.cc#L156
looks like a plausible cause for that.

Good ideas how to move on?  A quick&dirty solution would be to avoid postfix "Filter" here
https://github.com/cms-sw/cmssw/pull/16792/files#diff-5b29e53ee0c1d396908444c2711d39f6R114
but e.g. "Filtr" (or "filter" with lowecase "f"), and leave the validation logic to be enhanced by the authors (assuming HLT prefers to possibly use the "Filter" as part of the pixel tracking filters in the module labels). Or leave it as it is? (I'll anyway run a test overnight changing the postfix to something else, to confirm the diagnosis if nothing else)

(personally I feel checking only the module labels and not taking into account the C++ types in the HLT path "parsing" is fragile...)

**Comment by slava77** [comment]
On 12/8/16 1:26 PM, Matti Kortelainen wrote:
> Actually this might be related to this PR. |hltPixelTracksL3MuonFilter|
> is a module added in |customizeHLTforCMSSW| of type
> |PixelTrackFilterByKinematicsProducer|, whereas
> |hltMu7p5Track2UpsilonTrackMassFiltered| is of type
> |HLTMuonTrackMassFilter|. I wonder how this change enters up in a DQM
> histogram though. Ok, taking a quick look on the code, this logic
> https://github.com/cms-sw/cmssw/blob/CMSSW_9_0_X/HLTriggerOffline/HeavyFlavor/src/HeavyFlavorValidation.cc#L156
> looks like a plausible cause for that.

I removed my +1


> 
> Good ideas how to move on? A quick&dirty solution would be to avoid
> postfix "Filter" here

I think it's better to add a check of the module type in the
HeavyFlavorValidation

		--slava

**Comment by makortel** [comment]
> >  Good ideas how to move on? A quick&dirty solution would be to avoid
> >  postfix "Filter" here
> 
> I think it's better to add a check of the module type in the
> HeavyFlavorValidation

I have no idea what other type of modules the `HeavyFlavorValidation` would like to monitor than `HLTMuonTrackMassFilter`, so without more information I don't know how exactly to fix.

**Comment by makortel** [comment]
Or then I make it to ignore all the added `PixelTrackFIlter` producers.

**Comment by slava77** [comment]
On 12/8/16 2:13 PM, Matti Kortelainen wrote:
>         Good ideas how to move on? A quick&dirty solution would be to avoid
>         postfix "Filter" here
> 
>     I think it's better to add a check of the module type in the
>     HeavyFlavorValidation
> 
> I have no idea what other type of modules the |HeavyFlavorValidation|
> would like to monitor than |HLTMuonTrackMassFilter|, so without more
> information I don't know how exactly to fix.

"EDFilter"  == moduleEDMType(module)

this is used in a bunch of other DQM modules.

**Comment by makortel** [comment]
@cmsbuild, please test

**Comment by makortel** [comment]
@slava77 
> "EDFilter"  == moduleEDMType(module)

That seemed to do the trick (and to be generic-enough), thanks.

**Comment by slava77** [comment]
@makortel @boudoul @veszpv 
any idea why data (1000.0 and 136.731) MechanicalView plots go in and out at least in recent IBs?

This PR baseline vs PR bring the plots in:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_9_0_X_2016-12-08-2300+16792/17469/1000.0_RunMinBias2011A+RunMinBias2011A+TIER0+SKIMD+HARVESTDfst2+ALCASPLIT/AlCaReco_SiStrip.html

Unrelated to tracking PR #16909, same as this PR:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_9_0_X_2016-12-08-2300+16909/17480/1000.0_RunMinBias2011A+RunMinBias2011A+TIER0+SKIMD+HARVESTDfst2+ALCASPLIT/AlCaReco_SiStrip.html

Some other unrelated PR (tested in the same IB) shows no diffs
https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_9_0_X_2016-12-08-2300+16943/17473/1000.0_RunMinBias2011A+RunMinBias2011A+TIER0+SKIMD+HARVESTDfst2+ALCASPLIT/AlCaReco_SiStrip.html

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/16792 5e6513c
- the issue found in HLT validation after the last signoff  https://github.com/cms-sw/cmssw/pull/16792#issuecomment-265785056  was addressed
- jenkins tests pass and comparisons with baseline show no relevant differences (the diff in Upsilon plots in HLT is gone)

**Comment by slava77** [comment]
unhold

**Comment by slava77** [comment]
ping
we need DQM, HLT, AlCa and FastSim signatures here

**Comment by Martin-Grunewald** [comment]
please test
(HLT menu has changed since in 90X)

**Comment by Martin-Grunewald** [comment]
+1

**Comment by mmusich** [comment]
+1

**Comment by makortel** [comment]
For the follow-up PR I'll wait until either #17041 or #17042 gets merged (because I have to migrate the customize function, and doing anything there now would just create conflicts).

**Comment by fwyzard** [comment]
By the way @makortel , which approach do you prefer,  #17041 or #17042 ?
I'm trying to figure out with @Martin-Grunewald and @silviodonato which one to choose, feel free to chime in.

**Comment by makortel** [comment]
@fwyzard I can certainly share my experience from offline tracking, where are you having the discussion?(this PR is likely not the right forum)

**Pull request merged**